src/home_robot/home_robot/mapping/semantic/categorical_2d_semantic_map_module.py

```diff
@@ -577,6 +586,196 @@ def _update_local_map_and_pose(
+            t_p : t_p + x2 - x1, l_p : l_p + y2 - y1
+        ]
+        # Update the global map with the associated instances from the local map
+        global_instances_in_local = np.vectorize(instance_mapping.get)(
+            local_instances.cpu().numpy()
+        )
+        global_instances[x1:x2, y1:y2] = torch.tensor(
+            global_instances_in_local, dtype=torch.int64, device=global_instances.device
+        )
+        return global_instances
+
+    def get_instance_mapping(
+        self, extended_local_labels: Tensor, global_instances_within_local: Tensor
+    ) -> dict:
+        """
+        Creates a mapping of local instance IDs to global instance IDs.
+
+        Args:
+            extended_local_labels (Tensor): Labels of instances in the extended local map.
+            global_instances_within_local (Tensor): Instances from the global map within the local map's region.
+
+        Returns:
+            dict: A mapping of local instance IDs to global instance IDs.
+        """
+        instance_mapping = {}
+        max_instance_id = torch.max(global_instances_within_local)
+
+        # Associate instances in the local map with corresponding instances in the global map
+        for local_instance_id in torch.unique(extended_local_labels):
+            if local_instance_id == 0:
+                # ignore 0 as it does not correspond to an instance
+                continue
+
+            # pixels corresponding to
+            local_instance_pixels = extended_local_labels == local_instance_id
+
+            # Check for overlapping instances in the global map
+            overlapping_instances = global_instances_within_local[local_instance_pixels]
+            unique_overlapping_instances = torch.unique(overlapping_instances)
+
+            unique_overlapping_instances = unique_overlapping_instances[
+                unique_overlapping_instances != 0
+            ]
+
+            if len(unique_overlapping_instances) >= 1:
+                # If there is a corresponding instance in the global map, pick the first one and associate it
+                global_instance_id = unique_overlapping_instances[0].item()
+                instance_mapping[local_instance_id.item()] = global_instance_id
+            else:
+                # If there are no corresponding instances, create a new instance
+                new_global_instance_id = max_instance_id + 1
+                instance_mapping[local_instance_id.item()] = new_global_instance_id
+                max_instance_id += 1
+        instance_mapping[0] = 0
+        return instance_mapping
+
+    def add_instance_ids_to_global_map(
+        self, e: int, global_map: Tensor, local_map: Tensor, lmb: Tensor
+    ) -> Tensor:
+        """
+        Adds instance IDs to the global map based on the local map.
+
+        Args:
+            e (int): The index of the environment.
+            global_map (Tensor): The global map tensor.
```

Comment:
this is actually not a great description - assuming global map is the entire world map as a 2d map with channels representing object values?